### ISSUE TITLE
Plan for the core: Removing Qt, Decoupling, and Modernization

### ISSUE THREAD

**Issue by sakertooth** [LMMS/lmms#6472]
**Nothing about this is definitive, and not everything will be agreed upon. Feel free to tweak the following plan and give general feedback as you see fit.**

I am writing this plan as somewhat of a proposal as to what steps should be taken pertaining to the core, while also gathering the already existing ideas about it into one place. From a personal standpoint, improving the core has been an idea I have had for awhile (mainly because of the need for good foundations for potentially upcoming PRs), but that idea has been unclear in terms of how it should be approached. As a result, I have refrained from tackling it until I had a clear and strong plan to go forward. The hope is that this plan can be used as some ground to work on and develop ideas from.

## Preliminary Step: Working in a distinct branch away from master
This goal of drastically improving the core will most likely be far too large to start doing real work in the master branch. I fear it will also be polluted at some point given the sheer number of changes to be made, and the chance for merge conflicts are high. Instead, it can be done on its own branch, away from master.

### Reasoning
* It allows for easy migration to various frontends - We can use the core as a solid foundation for various frontends of our choosing, giving the application more flexibility in terms of how it can be improved.

* It will help assist in making the core realtime safe - [Qt may not be realtime safe](https://github.com/LMMS/lmms/wiki/Realtime-conventions#qt-data-structures), so using it in places where realtime is necessary makes Qt a liability.

* It will give space for more effective and efficient replacements to take effect - Removing Qt will allow more space for libraries such as the STL to shine through, [which should be the preferred option above other libraries when possible](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es1-prefer-the-standard-library-to-other-libraries-and-to-handcrafted-code).

* Developers wanting to contribute don't require knowing Qt beforehand - We will most likely still be using Qt as the default and standard frontend, but for those wanting to improve the internals of LMMS specifically, removing Qt may make this achieving this easier.

### Approach
I propose replacing simple Qt components and data structures first, such as:

* ``QVector``
* ``QMap``
* ``QHash``
* ``QPair``

Classes such as these already have drop-in replacements from the STL, so making these changes are fairly doable. A specific and detailed way of removing these classes, say, by file, is unwanted.

Functions such as ``qMin``, ``qMax``, ``qBound``, and ``qAbs`` also apply here.

### Missing functionality
Qt, being as vast as it is, comes with certain functionality used widely in this codebase that doesn't have a quick and easy drop-in replacement like the STL. The general approach to address this would be to prefer using a library before writing the functionality ourselves. If the situation calls for the opposite, by all means.

The functionality that will be missing after removing Qt and will need a suitable replacement are:
* Signals and Slots
* XML processing and handling
* Base64 Encoding and Decoding

## Goal: Decoupling the core from the GUI
Separate the core from the GUI.

### Reasoning

* Efficiency. It may be easier to accomplish the goals described here when the core is decoupled from the GUI because of the contained workflow this will create.
* Making the core as its own standalone library away from the frontend will be more doable.
* A more organized repository in general, which also ties in with better productivity and development.

### Approach 
The core can be organized in its own directory. More specifically, the core headers can be in ``include/core``, distinct from the GUI headers. [Do note that updating the core ``#include``'s in the source files have already been done, but reviewing is still needed.](https://github.com/Veratil/lmms/pull/7)

## Goal: Modernization
It is one thing to remove Qt and make the core its own library, but I feel it is necessary to also modernize the core as well.

### Reasoning
* The core currently uses old practices that were once deemed adequate, which can leave room for mistakes, memory leaks, and bugs.
* If we are removing Qt from the core and using a more idomatic alternative provided in the STL, it then gives reason to make the code idomatic as well on a larger scale.
* Having a strong and stable "core" (pun somewhat intended) to depend on is preferable when having to ponder about problems on the GUI side of things.  
* Idiomatic code is generally a good idea.

### Approach
Modernization of the core can be done in a fashion somewhat similar to clang-tidy checks, or done while removing Qt given a certain commit.

For instance, replacing the use of ``begin()`` and ``end()`` iterators in for-loops with a range-based for loop provided in C++11 if applicable. Such a change may be done once throughout the codebase, or simultaneously when Qt is being removed. These changes should not necessarily be bug fixes, but just changes that make the core more modern and idomatic.

## PR Format:
Given that the number of changes will most likely be fairly high, I propose an additional PR title format that can be followed. The reason behind this is that without some format to go off of, it can make it harder to keep track of the task and thus make everything less efficient.

Given a PR that is contributing to one of the goals desribed in this document, a title that can be used in the PR to make it easier to keep track of them can be:

"Removing Qt: [Goal], [Specific Details]"

Starting with "Removing Qt" makes it already clear that this PR in particular is for this project, and the goal that this PR is contributing towards is also specified. Additionally, specific details about the PR should be provided as per usual.

**Comment by irrenhaus3** [comment]
Mentioning #6452 as introducing another QT-provided type into the core: `QUuid`; add this to the "missing functionality" section.

**Comment by musikBear** [comment]
Just as a reminder:
https://github.com/LMMS/lmms/issues/6194

**Comment by Veratil** [comment]
> Just as a reminder: #6194

This has nothing to do with core though.

**Comment by sakertooth** [comment]
Added a "Methodology" section that explains the specific technical methods to alter and modify the code. The current verdict I have in summary is to both incrementally refactor the code while also having some clear canvas to start with. I find it hard to believe that one method is solely better than the other, especially when it is so dependent on the situation and the kind of code we are dealing with in the core.

**Comment by musikBear** [comment]
> This has nothing to do with core though.

Oh would we _keep_ QT for UI even if/when we outfaced it in core!
I diddent realize that was in the plan -Sorry. I hoped for a complete exit from QT.

**Comment by Veratil** [comment]
> > This has nothing to do with core though.
> 
> Oh would we _keep_ QT for UI even if/when we outfaced it in core! I diddent realize that was in the plan -Sorry. I hoped for a complete exit from QT.

Not necessarily. A frontend redesign though has nothing to do with the backend core. As we progress with this plan, we could have a Qt frontend, a Godot, a mobile, and whatever else anyone decides to make for a frontend.

**Comment by sakertooth** [comment]
Updated my methodology/approach. I think I may be able to officially start now or somewhere in a couple days give or take if I have time.

**Comment by gnudles** [comment]
I think we should prefer Base87 instead of Base64. It is much more compact, and saves a lot of space.

**Comment by Veratil** [comment]
> I think we should prefer Base87 instead of Base64. It is much more compact, and saves a lot of space.

I don't know much, nor can I find much, about Base87, but I am aware of Ascii85. Is this what you meant?

**Comment by gnudles** [comment]
@Veratil 
Yeah, my bad
